subscriber/common/sink/sink.go

```diff
@@ -0,0 +1,123 @@
+	Shutdown()
+}
+
+// Destination contains the table and columns that each job is storing data into
+// also records the behavior when encountering key errors
+type Destination struct {
+	// Table is table name
+	Table string
+	// ColumnNames are the list of column names after sorted
+	ColumnNames []string
+	// PrimaryKeys maps primary key columnName to its columnID after sorted
+	PrimaryKeys map[string]int
+	// PrimaryKeysInSchema maps primary key columnName to its columnID defined in schema
+	PrimaryKeysInSchema map[string]int
+	// AresUpdateModes defines update modes
+	AresUpdateModes []memCom.ColumnUpdateMode
+	// NumShards is the number of shards in the aresDB cluster
+	NumShards uint32
+}
+
+//NewAresDatabase(
+//	serviceConfig config.ServiceConfig, jobConfig *rules.JobConfig, cluster string,
+//	config client.ConnectorConfig) (Sink, error)
+//NewKafkaPublisher(jobConfig *rules.JobConfig, serviceConfig config.ServiceConfig, cluster string, kpCfg config.KafkaProducerConfig) (Sink, error)
+func Sharding(rows []client.Row, destination Destination, jobConfig *rules.JobConfig) map[uint32][]client.Row {
+	if destination.NumShards == 0 || destination.NumShards == 1 {
+		// in this case, there is no sharding in this aresDB cluster
+		return nil
+	}
+
+	shards := make(map[uint32][]client.Row)
+	for i := uint32(0); i < destination.NumShards; i++ {
+		shards[i] = make([]client.Row, 0, len(rows))
+	}
+
+	for _, row := range rows {
+		// convert primaryKey to byte array
+		pk := make([]byte, jobConfig.GetPrimaryKeyBytes())
+		err := GetPrimaryKeyBytes(row, destination, jobConfig, pk)
+		if err != nil {
+			utils.StackError(err, "Failed to convert primaryKey to byte array for row: %v", row)
+			continue
+		}
+
+		// calculate shard
+		shardID := shardFn(pk, destination.NumShards)
+		shards[shardID] = append(shards[shardID], row)
+	}
+	return shards
+}
+
+func shardFn(key []byte, numShards uint32) uint32 {
+	return utils.Murmur3Sum32(unsafe.Pointer(&key[0]), len(key), 0) % numShards
+}
+
+func GetPrimaryKeyBytes(row client.Row, destination Destination, jobConfig *rules.JobConfig, key []byte) error {
+	primaryKeyValues := make([]memCom.DataValue, len(destination.PrimaryKeys))
+	var err error
+	i := 0
+	for columnName, columnID := range destination.PrimaryKeys {
+		columnIDInSchema := destination.PrimaryKeysInSchema[columnName]
+		primaryKeyValues[i], err = GetDataValue(row[columnID], columnIDInSchema, jobConfig)
+		if err != nil {
+			return utils.StackError(err, "Failed to read primary key at row %d, col %d",
+				row, columnID)
+		}
+		i++
+	}
+
+	if err := memstore.GetPrimaryKeyBytes(primaryKeyValues, key); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetDataValue returns the DataValue for the given column value.
+func GetDataValue(col interface{}, columnIDInSchema int, jobConfig *rules.JobConfig) (memCom.DataValue, error) {
+	var dataStr string
+	var ok bool
+	if dataStr, ok = col.(string); !ok {
```

Comment:
is this col val after enum translation?

Reply:
this func is called only if datatype is non string